src/git.py

```diff
@@ -0,0 +1,69 @@
+from subprocess import run, PIPE
+
+# The idea here is to prevent deploying into production a system that
+# is untagged and/or has a dirty git status (the latter is most
+# commonly triggered by a submodule being out of date - seen with the
+# backup module).
+#
+# We only want this on production really.  We detect this by looking
+# to see if 'backup' is selected (this will be false for all but
+# production).
+#
+# It would be easy enough to add a check to ensure that the tag
+# actually exists on the remote machine, but that requires
+#
+#   git ls-remote --tags {remote} refs/tags/{tag}
+#
+# and that *does* use the network which slows things down a little.
+# This is only an issue if people are tagging on the production
+# machine, so it would be good if that did not happen.
+def git_check(settings):
+    strict = settings["backup"]
+    def report(msg):
+        if strict:
+            raise Exception(msg)
+        else:
+            print("NOTE: " + msg)
+
+    sha = git_sha()
+    if not sha:
+        # This happens on teamcity
+        report("running outside of git tree")
+        return
+
+    is_clean = git_is_clean()
+
+
+    if not is_clean:
+        report("git status reports directory is unclean")
+
+    tag = git_get_tag(sha)
+    if not tag:
+        report("HEAD is not tagged")
+        tag = "<<UNTAGGED>>"
+
+    print("This is montagu {tag} ({sha})".format(tag = tag, sha = sha))
+
+def git_is_clean():
+    p = run(["git", "status", "-s"], stdout = PIPE, stderr = PIPE)
```

Comment:
You may wish to add `check=True`, which will raise an exception if git exits with a non-zero exit code.

Reply:
For some reason I thought that for unclean status git returned exit code > 0 but that is not actually the case.  I've added this here